### PR TITLE
MtgFinder 0.2.0 | Tailwind integration and card images rendered.

### PR DESCRIPTION
**Pull request by dsmerrikin**

### CODE REVIEW

File: components/Card/Card.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const Card = (props) => {
```
Comment (richpressler) [inline components/Card/Card.js:1]: TS-ify me Cap'n! (Make it a `.TS` file)

Not going to affect anything functionally, but good habit to keep things uniform across a codebase

Comment (richpressler) [reply components/Card/Card.js:1]: ^ Following that, it's generally recommended to give your `props` an interface, and spread the props in the parameters, like this:

```
interface MyComponentProps {
  prop1: string
  prop2: number
  prop3: boolean
}

const MyComponent = ({
  prop1,
  prop2,
  prop3
}: MyComponentProps) => {
```

This makes it so you don't have to have that line `const data = props.meta` because `meta` will already be available!

File: pages/api/allprintings/[filterByLegality].ts [hunk 2]
```diff
@@ -34,10 +35,15 @@ export default async function handler(
   const data = [];
   for (let [, set] of Object.entries(parsedData.data)) {
     for (let [, card] of Object.entries(set.cards)) {
-      if (card.legalities[selectedLegality] === "Legal") {
-        data.push(card);
-        if (data.length > 99) {
-          return res.json(data);
+      if (!_.find(data, card.name)) {
```
Comment (richpressler) [inline pages/api/allprintings/[filterByLegality].ts:38]: I think this is a leftover from your troubleshooting

File: components/Card/Card.js (hunk 1 above)
Comment (dsmerrikin) [reply components/Card/Card.js:1]: So cool! Thanks for the lessons! Loving this!

Comment (richpressler) [reply components/Card/Card.js:1]: My pleasure my man! For educational purposes: the reasoning for this isn't just because "it's the convention" or "it's in the spec" etc. This helps educate your (and your fellow devs') IDE's so when they're implementing that nifty component you made, they get type-ahead help as well as type-checking to prevent annoying useless JS errors of the past 🔥

File: pages/api/allprintings/index.ts [hunk 3]
```diff
@@ -1,22 +1,30 @@
 import { NextApiRequest, NextApiResponse } from "next";
 import fs from "fs";
+import http
```
Comment (richpressler) [inline pages/api/allprintings/index.ts:3]: nit: no semi here

I know this is a learning project, so this isn't critical to learning react and TS, but this is still important stuff. This would get called out in a PR review on a real project. Few pointers:

1. Always run whatever lint command(s) you have set up, if they're not pre-commit hooks. Next comes with `npm run lint` already set up (an alias for `next lint`).
2. Easier said than done, since we have to be on projects with varying styles, be uniform when writing
3. VSCode has an auto-format on save option. Google it up, I don't use it anymore for personal reasons but it may be useful

File: pages/index.js [hunk 4]
```diff
@@ -1,28 +1,30 @@
...
+import _ from "lodash";
 import "bootstrap/dist/css/bootstrap.css";
+import Card from "../components/Card/Card";
 
 //Write a fetcher function to wrap the native fetch function and return the result of a call to url in json format
 const fetcher = (url) => fetch(url).then((res) => res.json());
 
 export default function Index() {
   //Set up SWR to run the fetcher function when calling "/api/allprintings"
   //There are 3 possible states: (1) loading when data is null (2) ready when the data is returned (3) error when there was an error fetching the data
```
Comment (richpressler) [inline pages/index.js:12]: Digging these comments 👍

File: tsconfig.json [hunk 5]
```diff
@@ -22,9 +18,8 @@
   "include": [
     "next-env.d.ts",
     "**/*.ts",
-    "**/*.tsx"
+    "**/*.tsx",
+    "components/Card/Card.js"
```
Comment (richpressler) [inline tsconfig.json:22]: Once you convert `Card.js` => `Card.ts`, this line won't be needed

File: pages/index.js [hunk 6]
```diff
@@ -1,28 +1,30 @@
...
   if (!data) return <div>Loading...</div>;
   //Handle the ready state and display the result contained in the data object mapped to the structure of the json file
   return (
-    <div>
-      <ul>
-        {data.map((card) => (
-          <li>{card.name}</li>
-        ))}
-      </ul>
+    <div className='grid grid-cols-6 gap-2'>
```
Comment (richpressler) [inline pages/index.js:21]: nit: single quote used here, double many other places (and single quote others as well).

Always be consistent in double/single quote usage.

Comment (richpressler) [reply pages/index.js:21]: Just my two cents, single quote is pretty much the way to go nowadays.

I'm Hip!

**Comment by richpressler** [comment]
Overall, everything looks great! Good logic, good delegating of filtering to the backend (in a setup with a database, the DB would do the filtering. Especially when the data is at this scale), good React code, just a few things in the way of general TS and style

File: components/Card/Card.js (hunk 1 above)
Comment (patnealcodes) [reply components/Card/Card.js:1]: exactly what Richard said - it takes a bit of overhead to get comfy with, but once you do it's a language-changer

File: pages/index.js (hunk 6 above)
Comment (patnealcodes) [reply pages/index.js:21]: A good rule of thumb is to set up your own guidelines for which quotes are used when. For instance I:
- Use double quotes for attribute names
- Use template literal quotes (``) when interpolating a variable in a string
- Use single quotes for regular string literals in JS/TS
- Use double quotes when a string contains single quotes as part of the string itself, and vice versa

You can typically use whichever you prefer as long as, like Richard said, it's consistent - and also try and follow precedence of an existing codebase

**Comment by dsmerrikin** [comment]
@richpressler thanks so much for the feedback!

File: pages/index.js (hunk 6 above)
Comment (dsmerrikin) [reply pages/index.js:21]: Much appreciated @patnealcodes & @richpressler ! I've had my vs code setup to auto-format with Prettify on save, so I may need to adjust some settings.

File: pages/api/allprintings/[filterByLegality].ts [hunk 7]
```diff
@@ -25,16 +28,24 @@ export default async function handler(
     "standard",
     "vintage",
   ];
-  const selectedLegality = req.query.filterByLegality.toLowerCase();
+  const selectedLegality = req.query.filterByLegality.toString().toLowerCase();
```
Comment (brandonaaskov) [inline pages/api/allprintings/[filterByLegality].ts:31]: If `filterByLegality` is ever undefined, this will throw an error. Not sure if that's ever going to be the case, but you can get around it a couple of ways.

[Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)
```js
req.query.filterByLegality?.toString().toLowerCase()
```

[Lodash: _.toLower()](https://lodash.com/docs/4.17.15#toLower)
```js
_.toLower(undefined) // ''
_.toLower(55) // '55'
```

Basically lodash is just generally forgiving when you pass in falsey values like `undefined` or `null`.

File: pages/api/allprintings/[filterByLegality].ts [hunk 8]
```diff
@@ -25,16 +28,24 @@ export default async function handler(
...
+      const cardData: CardInterface = card;
+      const cardLegalities: CardInterface["legalities"] = cardData.legalities;
+      const cardName: CardInterface["name"] = cardData.name;
+      if (
+        cardLegalities[selectedLegality] === "Legal" &&
+        !_.find(data, { name: cardName })
+      ) {
         data.push(card);
         if (data.length > 99) {
           return res.json(data);
```
Comment (brandonaaskov) [inline pages/api/allprintings/[filterByLegality].ts:51]: When you have a variable outside of a loop, but you're updating that variable from inside the loop, it's usually a dead giveaway that you could be using reduce or map instead. Glancing at this, I think a reducer is the right fit here.

File: pages/api/allprintings/[filterByLegality].ts [hunk 9]
```diff
@@ -25,16 +28,24 @@ export default async function handler(
...
+    for (let [, card] of Object.entries(setCards)) {
+      const cardData: CardInterface = card;
+      const cardLegalities: CardInterface["legalities"] = cardData.legalities;
+      const cardName: CardInterface["name"] = cardData.name;
+      if (
+        cardLegalities[selectedLegality] === "Legal" &&
+        !_.find(data, { name: cardName })
+      ) {
         data.push(card);
         if (data.length > 99) {
```
Comment (brandonaaskov) [inline pages/api/allprintings/[filterByLegality].ts:50]: In this case if you have more than 100 cards you're returning that data and stopping there. Why is that? I can see not wanting to return too large of a set because it would crash your browser, but maybe.... 500 or 1,000 instead? Basically, I don't want to get bogged down in pre-optimizing or paging the API since this just needs to run locally, but I want to make sure you're not returning too few cards in the set that it doesn't make sense visually.

Comment (dsmerrikin) [reply pages/api/allprintings/[filterByLegality].ts:50]: Yeah it takes a long time to load if its a giant number, so lets meet in the middle and say 750.
